fix(outputs.iotdb): Handle paths that contain illegal characters

giovanni-bellini-argo opened the pull request:
## Summary
<!-- Mandatory
Explain here the why, the rationale and motivation, for the changes.
-->
Support for IoTDB ILLEGAL_PATH error when tags contains an illegal character, explained more in detail in issue.

## Checklist
<!-- Mandatory
Please confirm the following by replacing the space with an "x" between the []:
-->

- [x] No AI generated code was used in this PR

## Related issues
<!-- Mandatory
All PRs should resolve an issue, if one does not exist, please open one.
-->
resolves #14518

powersj reviewed (commented):
Thanks for the PR!

Looking at the [IoTBD](https://iotdb.apache.org/UserGuide/V0.13.x/Reference/Syntax-Conventions.html#identifiers) this does appear to be the suggested way of handling special characters in the path.

Can you break out the regex checking into a new function and please add tests for that function with both cases including special characters? My only concern with these types of changes is future changes down the road or corner cases :)

Thanks!

srebhan commented:
@giovanni-bellini-argo please also sign the CLA!

giovanni-bellini-argo commented:
!signed-cla

powersj reviewed (commented):
Can you confirm my understanding after playing with this a bit more: A path is made up of three elements:

* root - reserved word
* storage group
* child - one or more items separated by periods

The storage group can only have:

> The storage group name can only be characters, numbers and underscores.

I think this is what you are currently testing for. However, the child elements seem to allow a greater set of elements. This means the following are valid without escaping them.

> root.foo.3
> root.foo.()

Is that your understanding as well? I've been trying this with version 0.13.0, so that may play into this as well.

giovanni-bellini-argo commented:
We are testing on V1.2.2 and could not create child nodes with any other characters outside of underscores and alphanumeric characters

powersj commented:
Thanks!

> We are testing on V1.2.2 and could not create child nodes with any other characters outside of underscores and alphanumeric characters.

I am glad we are trying on different versions. Unfortunately, having different behavior on different versions does mean we would need to make this opt-in as well. We cannot be breaking existing users on versions that do not run into this.

How we typically handle this is add a config option, something like:

```toml
# Mode to sanitize tags
# By default, tags are not sanitized. Possible options:
#   1.2 - only allows alpha/numeric and underscores, and non-numeric values. Otherwise, 
#         backticks are applied
sanitize_tag = ""
```

Then if that value is set to "1.2", you can apply your sanitization function.

I would still like to see a specific test as well that only tests the sanitize function with a various test cases.

Does that make sense?

powersj commented:
I've also asked over in the iotdb repo for guidance on valid characters:

https://github.com/apache/iotdb/discussions/11840

giovanni-bellini-argo commented:
a list of updates:
1. added the option as per your request and made some researches: from version 1.x.x they completely changed the handling of nodes names which can apparently take only letters, numbers and underscores but can't be composed of only numbers.
2. made some more checks to throw an error if root (which is a reserved kayword) is used and if ` is misused
3. added the specific tests cases.

powersj commented:
Thanks for the updates and doing that research. Almost there!

1. I do not see your name in our CLA signed sheet. Can you please sign it?
2. Looks like there a couple in lint issues:

```s
plugins/outputs/iotdb/iotdb.go:277:28  gosimple  S1007: should use raw string (`...`) with regexp.Compile to avoid having to escape twice
plugins/outputs/iotdb/iotdb.go:310:5   revive    var-naming: don't use underscores in Go names; var tag_value should be tagValue
```

3. And a test issue with your new tests, specifically: `TestTagsHandling/treat_tags_as_device_IDs`

4. Please update the sample config and README with the new config option for `sanitize_tag`.

Happy to help with any of these updates, let me know if you do want assistence.

giovanni-bellini-argo commented:
as per the CLA i did sign it on google (first time doing this so i cheched out online and found most use the google db) i did also read your guide about contributions and followed the link to sign the CLA on your site but the form has some iessues:

![image](https://github.com/influxdata/telegraf/assets/139997828/251fd5b7-fb1c-4c9c-ba1b-4c5fc41aa388)

powersj commented:
Hey @giovanni-bellini-argo,

That seems to be your browser saying it is blocked by your content security policy. Can you please try a different browser?

Thanks!

powersj commented on plugins/outputs/iotdb/README.md:
```diff
@@ -141,4 +141,19 @@ to use them.
+  ## Handling of Unsupported Characters
+  ## Some characters in different versions of IoTDB are not supported in path name
+  ## A guide with suggetions on valid paths can be found here:
+  ## for iotdb 0.13.x           -> https://iotdb.apache.org/UserGuide/V0.13.x/Reference/Syntax-Conventions.html#timestamp-literals
```
Should this link point to the identifiers instead:

https://iotdb.apache.org/UserGuide/V0.13.x/Reference/Syntax-Conventions.html#identifiers

powersj commented on plugins/outputs/iotdb/iotdb.go:
```diff
@@ -229,6 +231,60 @@ func (s *IoTDB) convertMetricsToRecordsWithTags(metrics []telegraf.Metric) (*rec
+
+	// IoTDB uses "root" as a keyword and can be called only at the start of the path
+	if tag == "root" {
+		return "", fmt.Errorf("can't append keyword \"root\" to IoTDB base path")
```
Should this error message actually read that you "cannot use \"root\" as tag?

```suggestion
		return "", errors.New("cannot append keyword \"root\" to IoTDB base path")
```

powersj commented on plugins/outputs/iotdb/iotdb.go:
```diff
@@ -229,6 +231,60 @@ func (s *IoTDB) convertMetricsToRecordsWithTags(metrics []telegraf.Metric) (*rec
+	if tag == "root" {
+		return "", fmt.Errorf("can't append keyword \"root\" to IoTDB base path")
+	} else if forbiddenBacktick.MatchString(tag) { // returns an error if the backsticks are used in an inappropriate way
+		return "", fmt.Errorf("can't use ` in IoTDB base path")
```
```suggestion
		return "", errors.New("cannot use ` in IoTDB base path")
```

powersj commented on plugins/outputs/iotdb/iotdb.go:
```diff
@@ -251,7 +307,13 @@ func (s *IoTDB) modifyRecordsWithTags(rwt *recordsWithTags) error {
 			topic := []string{rwt.DeviceIDList[index]}
 			for _, tag := range tags { // for each tag, append it's Value
-				topic = append(topic, tag.Value)
+				tagValue, err := s.validateTag(tag.Value) // validates tag
+
+				if err != nil {
+					return err
+				}
+
+				topic = append(topic, tagValue)
```
```suggestion
				tagValue, err := s.validateTag(tag.Value) // validates tag
				if err != nil {
					return err
				}
				topic = append(topic, tagValue)
```

powersj commented on plugins/outputs/iotdb/sample.conf:
```diff
@@ -43,3 +43,18 @@
+  ## Handling of Unsupported Characters
+  ## Some characters in different versions of IoTDB are not supported in path name
+  ## A guide with suggetions on valid paths can be found here:
+  ## for iotdb 0.13.x           -> https://iotdb.apache.org/UserGuide/V0.13.x/Reference/Syntax-Conventions.html#timestamp-literals
```
Same question here about the URL anchor.

powersj commented on plugins/outputs/iotdb/iotdb_test.go:
```diff
@@ -271,6 +271,68 @@ func TestMetricConversionToRecordsWithTags(t *testing.T) {
+			plugin:   func() *IoTDB { s := newIoTDB(); s.SanitizeTags = "1.3"; return s }(),
+			expected: []string{"", ""},
+			input:    []string{"root", "wo`rd"},
+		},
+
+		{
```
```suggestion
		},
		{
```

powersj reviewed (commented):
Sorry for the delay, thanks again for the re-work, I think we are nearly there!

powersj reviewed (approved):
Thanks for working through this!

srebhan commented on plugins/outputs/iotdb/README.md:
```diff
@@ -141,4 +141,19 @@ to use them.
   ##   - "device_id"  --  root.sg.device.private.working, s1=100, s2="hello"
   # convert_tags_to = "device_id"
+
+  ## Handling of Unsupported Characters
```
```suggestion
  ## Handling of unsupported characters
```

srebhan commented on plugins/outputs/iotdb/iotdb.go:
```diff
@@ -229,6 +231,60 @@ func (s *IoTDB) convertMetricsToRecordsWithTags(metrics []telegraf.Metric) (*rec
+	// matches any word that has a non valid backtick
+	// `word`  							 <- dosen't match
+	// ``word , `wo`rd` , `word , word`  <- match
+	forbiddenBacktick, _ := regexp.Compile("^[^\x60].*?[\x60]+.*?[^\x60]$|^[\x60].*[\x60]+.*[\x60]$|^[\x60]+.*[^\x60]$|^[^\x60].*[\x60]+$")
+	allowedBacktick, _ := regexp.Compile("^[\x60].*[\x60]$")
```
If we know this should compile, you should do
```suggestion
	forbiddenBacktick := regexp.MustCompile("^[^\x60].*?[\x60]+.*?[^\x60]$|^[\x60].*[\x60]+.*[\x60]$|^[\x60]+.*[^\x60]$|^[^\x60].*[\x60]+$")
	allowedBacktick := regexp.MustCompile("^[\x60].*[\x60]$")
```
to catch programming errors. We might also want define those two as package-global variables to not compile them for every metric again as they seem constant...

srebhan commented on plugins/outputs/iotdb/iotdb.go:
```diff
@@ -229,6 +231,60 @@ func (s *IoTDB) convertMetricsToRecordsWithTags(metrics []telegraf.Metric) (*rec
+
+	// IoTDB uses "root" as a keyword and can be called only at the start of the path
+	if tag == "root" {
+		return "", errors.New("cannot use \"root\" as tag")
```
```suggestion
		return "", errors.New("cannot use 'root' as tag")
```

srebhan commented on plugins/outputs/iotdb/iotdb.go:
```diff
@@ -229,6 +231,60 @@ func (s *IoTDB) convertMetricsToRecordsWithTags(metrics []telegraf.Metric) (*rec
+
+	switch s.SanitizeTags {
+	case "0.13":
+		matchUnsopportedCharacter, _ := regexp.Compile("[^0-9a-zA-Z_:@#${}\x60]")
+
+		regex := []*regexp.Regexp{matchUnsopportedCharacter}
+		regexArray = append(regexArray, regex...)
+
+	// from version 1.x.x IoTDB changed the allowed keys in nodes
+	case "1.0", "1.1", "1.2", "1.3":
+		matchUnsopportedCharacter, _ := regexp.Compile("[^0-9a-zA-Z_\x60]")
+		matchNumericString, _ := regexp.Compile(`^\d+$`)
```
Same here, use `MustCompile` and initialize `regexArray` in `Init()`!

giovanni-bellini-argo replied:
why initializing `regexArray` in `Init()`?
u think it would be better to include it in the struct for future changes?

and i would also delete those useless variables and directly append the regex to the array.
```golang
regexArray := []*regexp.Regexp{} // array of compiled regex patterns

	switch s.SanitizeTags {
	case "0.13":
		regex := []*regexp.Regexp{
			regexp.MustCompile("[^0-9a-zA-Z_:@#${}\x60]"),
		}
		regexArray = append(regexArray, regex...)

	// from version 1.x.x IoTDB changed the allowed keys in nodes
	case "1.0", "1.1", "1.2", "1.3":
		regex := []*regexp.Regexp{
			regexp.MustCompile("[^0-9a-zA-Z_\x60]"),
			regexp.MustCompile(`^\d+$`),
		}

		regexArray = append(regexArray, regex...)
	default:
		return tag, nil
	}
```
the idea is to make it easy to undertand and modify

srebhan commented on plugins/outputs/iotdb/sample.conf:
```diff
@@ -43,3 +43,18 @@
   ##   - "device_id"  --  root.sg.device.private.working, s1=100, s2="hello"
   # convert_tags_to = "device_id"
+
+  ## Handling of Unsupported Characters
```
```suggestion
  ## Handling of unsupported characters
```

srebhan reviewed (commented):
Thanks for this nice PR @giovanni-bellini-argo! I have some small comments, which moving the regexp-compilation out of the `Write()` path being the most important one...

srebhan commented on plugins/outputs/iotdb/iotdb.go:
```diff
@@ -229,6 +231,60 @@ func (s *IoTDB) convertMetricsToRecordsWithTags(metrics []telegraf.Metric) (*rec
+		return tag, nil
+	}
+
+	regexArray := []*regexp.Regexp{} // array of compiled regex patterns
+
+	switch s.SanitizeTags {
+	case "0.13":
+		matchUnsopportedCharacter := regexp.MustCompile("[^0-9a-zA-Z_:@#${}\x60]")
+
+		regex := []*regexp.Regexp{matchUnsopportedCharacter}
+		regexArray = append(regexArray, regex...)
+
+	// from version 1.x.x IoTDB changed the allowed keys in nodes
+	case "1.0", "1.1", "1.2", "1.3":
+		matchUnsopportedCharacter := regexp.MustCompile("[^0-9a-zA-Z_\x60]")
+		matchNumericString := regexp.MustCompile(`^\d+$`)
+
+		regex := []*regexp.Regexp{matchUnsopportedCharacter, matchNumericString}
+		regexArray = append(regexArray, regex...)
+	default:
+		return tag, nil
+	}
```
Please move this part to the `Init()` function and store `regexArray` (maybe name it `sanityRegex`) in the `IoTDB` structure. This way, we don't need to do this for every metric we write but only once!

srebhan commented on plugins/outputs/iotdb/iotdb.go:
```diff
@@ -229,6 +231,60 @@ func (s *IoTDB) convertMetricsToRecordsWithTags(metrics []telegraf.Metric) (*rec
+	// matches any word that has a non valid backtick
+	// `word`  							 <- dosen't match
+	// ``word , `wo`rd` , `word , word`  <- match
+	forbiddenBacktick := regexp.MustCompile("^[^\x60].*?[\x60]+.*?[^\x60]$|^[\x60].*[\x60]+.*[\x60]$|^[\x60]+.*[^\x60]$|^[^\x60].*[\x60]+$")
+	allowedBacktick := regexp.MustCompile("^[\x60].*[\x60]$")
```
Please move this outside of any function, similar to how `sampleConfig` is declared. This way, we need to compile the **static** expressions only once and not for every metric!

srebhan commented on plugins/outputs/iotdb/iotdb.go:
```diff
@@ -229,6 +231,60 @@ func (s *IoTDB) convertMetricsToRecordsWithTags(metrics []telegraf.Metric) (*rec
 	return rwt, nil
 }
 
+// takes in an array of compiled regex patterns
+// and validates them on the given tag.
+func matchToEncloseInBacktick(regexStrings []*regexp.Regexp, tag string) string {
+	// loops through all the regex patterns and if one
+	// pattern matches returns the tag between `
+	for _, regex := range regexStrings {
+		if regex.MatchString(tag) {
+			return "`" + tag + "`"
+		}
+	}
+
+	return tag
+}
```
You should really merge this into `validateTag`...

srebhan reviewed (commented):
Thanks for your update @giovanni-bellini-argo! Please try to compile the regular-expressions only once as suggested in the code! If you need help, I can provide a patch...

srebhan commented on plugins/outputs/iotdb/iotdb.go:
```diff
@@ -30,9 +37,11 @@ type IoTDB struct {
 	Log             telegraf.Logger `toml:"-"`
 
-	session *client.Session
+	SanityRegex []*regexp.Regexp
```
Please do not export this field as it is (and should be) only used internally
```suggestion
	sanityRegex []*regexp.Regexp
```

srebhan commented on plugins/outputs/iotdb/iotdb.go:
```diff
@@ -74,6 +83,22 @@ func (s *IoTDB) Init() error {
+	case "0.13":
+		matchUnsopportedCharacter := regexp.MustCompile("[^0-9a-zA-Z_:@#${}\x60]")
+
+		regex := []*regexp.Regexp{matchUnsopportedCharacter}
+		s.SanityRegex = append(s.SanityRegex, regex...)
```
```suggestion
		s.sanityRegex = append(s.sanityRegex, matchUnsupportedCharacter)
```

srebhan commented on plugins/outputs/iotdb/iotdb.go:
```diff
@@ -74,6 +83,22 @@ func (s *IoTDB) Init() error {
+		matchUnsopportedCharacter := regexp.MustCompile("[^0-9a-zA-Z_\x60]")
+		matchNumericString := regexp.MustCompile(`^\d+$`)
+
+		regex := []*regexp.Regexp{matchUnsopportedCharacter, matchNumericString}
+		s.SanityRegex = append(s.SanityRegex, regex...)
```
```suggestion
		s.sanityRegex = append(s.sanityRegex, matchUnsupportedCharacter, matchNumericString)
```

srebhan commented on plugins/outputs/iotdb/iotdb.go:
```diff
@@ -74,6 +83,22 @@ func (s *IoTDB) Init() error {
+
+	// from version 1.x.x IoTDB changed the allowed keys in nodes
+	case "1.0", "1.1", "1.2", "1.3":
+		matchUnsopportedCharacter := regexp.MustCompile("[^0-9a-zA-Z_\x60]")
```
```suggestion
		matchUnsupportedCharacter := regexp.MustCompile("[^0-9a-zA-Z_\x60]")
```

srebhan commented on plugins/outputs/iotdb/iotdb.go:
```diff
@@ -74,6 +83,22 @@ func (s *IoTDB) Init() error {
 
+	switch s.SanitizeTags {
+	case "0.13":
+		matchUnsopportedCharacter := regexp.MustCompile("[^0-9a-zA-Z_:@#${}\x60]")
```
```suggestion
		matchUnsupportedCharacter := regexp.MustCompile("[^0-9a-zA-Z_:@#${}\x60]")
```

srebhan commented on plugins/outputs/iotdb/iotdb.go:
```diff
@@ -229,6 +254,28 @@ func (s *IoTDB) convertMetricsToRecordsWithTags(metrics []telegraf.Metric) (*rec
+
+	// loops through all the regex patterns and if one
+	// pattern matches returns the tag between `
+	for _, regex := range s.SanityRegex {
```
```suggestion
	for _, regex := range s.sanityRegex {
```

srebhan reviewed (commented):
Looks good @giovanni-bellini-argo! I would just un-export the member variable and fix the typos and we are good to go. Tried to do it in my suggestions to ease your life but no sure I got all of them... :-)

giovanni-bellini-argo commented:
done everything, thanks for the patience and help :sunglasses:

srebhan reviewed (approved):
Thank you for your work and patience with my requests @giovanni-bellini-argo! :-)
